### DISCUSSION TITLE
HomeAssistant tasmota integration and Berry driver as MQTT aggregator

### DISCUSSION THREAD

**Discussion by msetina** [arendst/Tasmota#21620] [General]
I am trying to implement a MQTT aggregator where sensors are constructed by calculating data from mqtt sensors. The problem arises when I expect to see those sensors in HomeAssistant. 
If I understand properly the discovery message is sent at startup from information added to JSON from berry driver. In this case the real sensor data is comes via MQTT to the device and probably the discovery is sent around the same time the subscriptions to those real sensors is made. Consequently the data from the device can be prepared after that so the discovery would never see that. 
Is there a way to repeat or delay discovery message?
In my case I could check if all sensors were reported and then run the discovery propagation.

**Comment by msetina** [comment]
OK. To anyone trying something like this. Be patient. If the driver is loaded with autoexec.be the discovery will be done. On the other hand in my case later also sensors were registered even if they were not present at start.

**Reply by sfromis** [reply]
The way the discovery for sensors work is that all the sensor drivers, including when coded in Berry, are called to produce the content of the `tasmota/discovery/(mac)/sensors` payload, without options to repeat or delay this.

This has the implication that the Berry driver has to be added at this time, and has to be able to produce a payload illustrating what sensor fields it is providing. If your Berry driver fails this, no discovery until next MQTT (re)connection, which thus indirectly is a repeat of the discovery messages.

**Reply by msetina** [reply]
But in my case it did add all the sensors that were "created" after (remote mqtt sensors report with 10s period) all have chimed in. What I can report that integration did discover sensors on the next sensor reporting time (5min after startup).
The payload you are talking about is fully available after all remote sensors chime in.
If I waited after maualy loading the driver it never happened. But when I added autoexec.be, sesors were discovered also after the next MQTT reporting period (5min).

**Reply by msetina** [reply]
It is a little indetermined. RIght now I see data from 3 out of 4 remote sensors. Probably some cutoff after startup.
This time can be determined in my case and if I could set the time before discovery would help the determinacy.

**Reply by sfromis** [reply]
What I'd do in such a case would be to have the aggregator driver (if you even need such a thing to republish the data), keep persistent data of what sensors are available, allowing it to publish a full model payload right after boot, before the "remote" sensors have chimed in.

**Reply by msetina** [reply]
My implementation right now is dynamic (we can have 1 phase or 3phase meters) also in data model, but current use can be forced in this direction if setup also adds meter type to be able to define data model.

**Reply by barbudor** [reply]
An interresting feature for that use case or similar (like BLE or Zigbee) that have dynamic sensors would a `tasmota` object method to re-trigger the discovery messages on-demand

(ping @s-hadinger)

**Reply by sfromis** [reply]
It could also be a command able to be triggered by other means than Berry, like scripting or maybe some type of change of what values are included from some driver.

IIRC there is some "ongoing" (not sure how actively) discussion behind the scenes about improving the discovery protocol around current limitations making it somewhat inflexible.

**Reply by msetina** [reply]
@sfromis Thank you. Your suggestion is exactly what is needed to mitigate current shortcomings. What @barbudor  suggested would help to have the same dynamics with current method of discovery. 
To extend the method of discovery maybe a file with types and/or sensors could be put on filesystem where the discovery process would find it. In case it is missing current method could be used.

**Reply by arendst** [reply]
What about the already available command ``Discover 1``?

**Reply by barbudor** [reply]
Oh s..t
I'd better read the docs then 🤣

**Reply by sfromis** [reply]
First, you'll have to put it into the docs :grin: 

It's basically a synonym of `SetOption19` except inverted. The non-inverted synonym is `DiscoverOff`. Hence commands `SetOption19 0` and `DiscoverOff 0` will do the same as `Discover 1`. Sum of this is "a tad confusing"....

**Reply by barbudor** [reply]
But what I understood from @arendst comment is that executing `Discovery 1` at any time when already enabled would trigger a re-send of the discovery messages

**Reply by arendst** [reply]
Many roads lead to Rome ....

**Reply by arendst** [reply]
@barbudor indeed.

EDIT: code
```
void CmndTasDiscover(void) {
  if (XdrvMailbox.payload >= 0) {
    Settings->flag.hass_discovery = !(XdrvMailbox.payload & 1);  // SetOption19 - Tasmota discovery (0)
    TasRediscover();
  }
  ResponseCmndChar(GetStateText(!Settings->flag.hass_discovery));
}
```

**Reply by msetina** [reply]
OK. So this is undocumented command whic would be usefull for my case. Can we expect Berry/tasmota namespace equivalent?

**Reply by sfromis** [reply]
`tasmota.cmd('Discover 1')` should work from Berry.
